Move vanadium processing onto focus tab of Engineering UI

DannyHindson opened the pull request:
**Description of work.**

The generation of the files required for Vanadium normalization is now done on the Focus tab of the user interface. This means the Vanadium data can be updated without having to rerun the Ceria calibration.

This PR includes a few other related changes:
- the setting "Force Vanadium Recalculation" has been removed and the Vanadium run number input has been moved from the Calibration tab to the Focus tab
- The Vanadium run number is now part of the processed vanadium workspace names in the ADS
- The Vanadium run number is also no longer written to the prm generated on the Calibration tab.

**To test:**

Open the engineering diffraction user interface (Interfaces, Diffraction menu)
Go to the Calibration tab
Run a calibration by selecting the "Create New Calibration" radio button and use Calibration Sample 193749
Go to the Focus tab
The region of interest should say "North and South banks"
Enter Vanadium 236516, Sample 299080
Hit Focus button

Check this all completes without errors

Fixes #31904. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

DanielMurphy22 committed:
Move Vanadium processing onto the Focus tab

Move the Vanadium run number edit box onto the Focus tab and
generate the processed Vanadium files as part of the Focus action
instead of them being prepared on the Calibration tab
Generation time is small enough that they can always be generated
from scratch now rather than saved out to file (although if they are
in the ADS from a previous focus they won't be regenerated)
Now that code checks if Vanadium ws in already in ADS, add
run number to workspace name to support scenario where user chooses
a different Vanadium run

DannyHindson committed:
Adjust comment now that V run not included

DannyHindson committed:
Add unit test for get_info_from_file

Previously it was just mocked so parsing logic wasn't tested

RichardWaiteSTFC commented on docs/source/release/v6.2.0/diffraction.rst:
```diff
@@ -54,6 +54,9 @@ New features
 - The last used RB number is now saved for the next session
+- The generation of the files required for Vanadium normalization is now done on the Focus tab of the user interface. This means the Vanadium data can be updated without
+having to rerun the Ceria calibration. As part of this change the setting "Force Vanadium Recalculation" has been removed and the Vanadium run number input has been
+moved from the Calibration tab to the Focus tab. The Vanadium run number is also no longer written to the prm generated on the Calibration tab
```
Should we mention in the release notes that this is a breaking change - i.e. old calibrations cannot be loaded (not a big deal as takes < 1 min to make a new one).

DannyHindson replied:
I've just tried loading an old .prm file and it doesn't work. Although not because it's trying to load the ceria run number from the file name. The problem occurs in the logic that tries to construct the file names for the calibration .nxs files that are also loaded along with the .prm file.
I think it might be a bit messy to make this backwards compatible because of the fact that the roi can sometimes contain an underscore. So can't really parse from the right hand end of the .prm name to strip off the roi. Instead we need to rely on the left hand section having a fixed number of _'s. So will add a comment to the release note that it's breaking.
While doing this I noticed that the exception handling around the file loading in the calibration model isn't v nice - lots of unhandled exceptions so I'll at least tidy that up

RichardWaiteSTFC commented on scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py:
```diff
@@ -42,16 +40,19 @@ def create_new_calibration(self,
-        except ValueError:
-            logger.error("Error loading Full instrument calibration - this is set in the interface settings.")
-            return
+        if Ads.doesExist("full_inst_calib"):
```
Nice to add check for `full_inst_calib` existing in ADS here

RichardWaiteSTFC reviewed (commented):
This is a big improvement, logically it makes sense to put the vanadium with the focussing and this negates the need to save the vanadium whole instrument (time-consuming) and allowed the unnecessary "Force vanadium recalc" option to be removed . Functionally everything works well, calibrations can be made and loaded and the vanadium run is remembered between sessions. 
One impact of the PR is that  old calibrations (probably just the filename) are no longer compatible. Do we want/is it easy to add some backward compatibility (e.g. checking if two run numbers exist in prm filename and if so taking the last one as the ceria)? If not I think just putting a comment in the release notes will suffice and I will approve (also we should remember to let Saurabh know when this PR is merged).

DannyHindson committed:
Couple of changes

Highlight breaking change in release note
Improve exception handling around loading of existing calibration

RichardWaiteSTFC commented on scripts/Engineering/gui/engineering_diffraction/tabs/calibration/model.py:
```diff
@@ -115,18 +115,28 @@ def extract_b2b_params(workspace):
 
     def load_existing_calibration_files(self, file_path):
+        import pydevd_pycharm
+        pydevd_pycharm.settrace(stdoutToServer = True, stderrToServer = True)
```
Looks like you left this debugging line in by accident!

DannyHindson replied:
Good spot! I've just removed it

DannyHindson commented:
Seems like my attempt to use mock_open doesn't work on Ubuntu\RHEL for some reason. It's not mocking the file properly and I can reproduce the problem on my Ubuntu VM. I'll have a look - but will remove if it's not straightforward

RichardWaiteSTFC commented:
> Seems like my attempt to use mock_open doesn't work on Ubuntu\RHEL for some reason. It's not mocking the file properly and I can reproduce the problem on my Ubuntu VM. I'll have a look - but will remove if it's not straightforward

Ok sounds reasonable I think the parsing of the file cold also be checked in a system test (in a non-mocked way)?

DannyHindson committed:
Change code to read prm file so it can be mocked

iterating through the object returns by mock_open only
seems to work on python 3.8 and not on 3.6. Change the
code to call readlines and iterate through that instead

This isn't functionally exactly the same because the whole
file is read in at once but the prm files aren't massive
so should be OK

RichardWaiteSTFC reviewed (approved):
Changes look good, did a quick test just to confirm and everything is OK. Tests passing as well so approving.

DannyHindson commented:
@mantidproject/gatekeepers - please could someone merge this? There's another PR that is dependent on these changes that we'd like to progress
